dattri/algorithm/influence_function.py

```diff
@@ -606,49 +606,231 @@ def __init__(
+        batch_size = test_rep.shape[0]
+        for layer in range(layer_cnt):
+            grad_layer = full_rep_split[layer]
+            # length = full train data size
+            length = grad_layer.shape[0]
+            # Split gradients into smaller chunks
+            grad_batches = grad_layer.split(batch_size, dim=0)
+            running_contributions = 0
+            for batch in grad_batches:
+                reg = 0.1 if regularization is None else regularization
+                contribution = torch.func.vmap(
+                    lambda grad, layer=layer, reg=reg: _transform_single_test_rep(
+                    query_split[layer],
+                    grad,
+                    reg,
+                ),
+                )(batch)
+                # Sum up the contributions by chunk and average at the end
+                running_contributions += contribution.sum(dim=0)
+            running_contributions /= length
+            ifvps.append(running_contributions)
+        return torch.cat(ifvps, dim=1)
+
+    def get_layer_wise_reps(self,
+        ckpt_idx: int,
+        query: torch.Tensor) -> Tuple[torch.Tensor, ...]:
+        """Split a representation into layer-wise representations.
 
-        self.ihvp_func = ifvp_datainf(
-            self.task.get_loss_func(),
-            0,
-            (None, 0),
-            param_layer_map=param_layer_map,
-            **self.transformation_kwargs,
-        )
+        Args:
+            ckpt_idx (int): The index of the model parameters. This index
+                is used for ensembling of different trained model.
+            query (torch.Tensor): Input representations to split, could be
+                train/test representations, of shape (batch_size,parameter)
 
+        Returns:
+            Tuple[torch.Tensor, ...]: The layer-wise splitted tensor, a tuple of shape
+                (batch_size,layer0_size), (batch_size,layer1_size)...
+        """
+        model_params, param_layer_map = self.task.get_param(
+            ckpt_idx, layer_split=True,
+        )
         split_index = [0] * (param_layer_map[-1] + 1)
         for idx, layer_index in enumerate(param_layer_map):
             split_index[layer_index] += model_params[idx].shape[0]
-
         current_idx = 0
         query_split = []
         for i in range(len(split_index)):
-            query_split.append(test_rep[:, current_idx : split_index[i] + current_idx])
+            query_split.append(query[:, current_idx : split_index[i] + current_idx])
             current_idx += split_index[i]
+        return query_split
 
-        vector_product = 0
-        for full_data_ in self.full_train_dataloader:
-            # move to device
-            full_data = tuple(data.to(self.device) for data in full_data_)
-            res = self.ihvp_func(
-                (model_params, (full_data[0], full_data[1].view(-1, 1).float())),
-                query_split,
+    def attribute(
```

Comment:
Why can't we just change the functions for generating and transforming representations and reuse the original attribute function?

Reply:
This is because I need to calculate full training data gradient before I start calculating transformed queries. Also this should be done only once for all batches of training and testing data. If I were to put this in one of the transformations the gradients of full_training_data is taken multiple times and leads to slow performance.

Reply:
It should be done in .cache then. See the Arnoldi attributor for example.

Reply:
Thanks. Will follow Arnoldi for caching.